File: java/src/com/mogreet/mercury/user/Transactions.java
/**
* Copyright 2011 Nicolas Payot
*
* This file is part of Mercury.
*
* Mercury is free software: you can redistribute it and/or modify
* it under the terms of the GNU General Public License as published by
* the Free Software Foundation, either version 3 of the License, or
* (at your option) any later version.
*
* Mercury is distributed in the hope that it will be useful,
* but WITHOUT ANY WARRANTY; without even the implied warranty of
* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
* GNU General Public License for more details.
*
* You should have received a copy of the GNU General Public License
* along with Mercury.  If not, see <http://www.gnu.org/licenses/>.
*/

package com.mogreet.mercury.user;

import com.mogreet.mercury.*;
import java.util.*;
import org.w3c.dom.*;
import javax.xml.xpath.*;

/**
* The Transactions object contains the response from a {@link Mercury#transactions} request to the Moms API.
*/
public class Transactions extends Response {

	private HashMap<String, HashMap<Integer, HashMap<String, String>>> campaigns;

	/**
	* Creates a new Transactions response object.
	* @param xmlDoc XML DOM document that is returned by all the requests.
	*/
	public Transactions(Document xmlDoc) {
		super(xmlDoc);
		if (!super.responseIsValid())
			throw new RuntimeException(super.getMessage());
		try {
			XPath xpath = XPathFactory.newInstance().newXPath();
			XPathExpression expr;
			String tmp = "/response/campaign/@id | /response/campaign/@name | ";
			tmp += "/response/campaign/transaction/@datestamp | /response/campaign/transaction/@hash | /response/campaign/transaction/@message_id | ";
			tmp += "/response/campaign/transaction/to | /response/campaign/transaction/from";
			expr = xpath.compile(tmp);
			NodeList nodes = (NodeList)(expr.evaluate(xmlDoc, XPathConstants.NODESET));
			
			campaigns = new HashMap<String, HashMap<Integer, HashMap<String, String>>>();
			String campaignsIdName = null;
			String datestamp = null;
			String hash = null;
			HashMap<String, String> transactionsInfo = null;
			
			for (int i = 0; i < nodes.getLength(); ++i) {
				if (nodes.item(i).getNodeName().equals("id")) {
					campaignsIdName = nodes.item(i).getNodeValue();
				} else if (nodes.item(i).getNodeName().equals("name")) {
					campaignsIdName += " " + nodes.item(i).getNodeValue();
					if (campaigns.get(campaignsIdName) == null)
						campaigns.put(campaignsIdName, new HashMap<Integer, HashMap<String, String>>());
				} else if (nodes.item(i).getNodeName().equals("datestamp")) {
					datestamp = nodes.item(i).getNodeValue();
				} else if (nodes.item(i).getNodeName().equals("hash")) {
					hash = nodes.item(i).getNodeValue();
				} else if (nodes.item(i).getNodeName().equals("message_id")) {
					HashMap<Integer, HashMap<String, String>> transactions = campaigns.get(campaignsIdName);
					transactions.put(Integer.parseInt(nodes.item(i).getNodeValue()), new HashMap<String, String>());
					transactionsInfo = transactions.get(Integer.parseInt(nodes.item(i).getNodeValue()));
					transactionsInfo.put("datestamp", datestamp);
					transactionsInfo.put("hash", hash);
				} else if (nodes.item(i).getNodeName().equals("to")) {
					transactionsInfo.put("to_number", ((Attr)nodes.item(i).getAttributes().item(0)).getValue());
					transactionsInfo.put("to_name", nodes.item(i).getFirstChild().getNodeValue());
				} else if (nodes.item(i).getNodeName().equals("from")) {
					transactionsInfo.put("from_number", ((Attr)nodes.item(i).getAttributes().item(0)).getValue());
					transactionsInfo.put("from_name", nodes.item(i).getFirstChild().getNodeValue());
				}
			}
		} catch (XPathExpressionException e) {
			throw new RuntimeException("\nAn error occured while parsing the XML data for the TRANSACTIONS call: " + e.getMessage());
		}
	}
	
	/**
	* @return The list of all the campaigns ID.
	*/
	public ArrayList<Integer> getCampaignsIdList() {
		ArrayList<Integer> campaignsIdList = new ArrayList<Integer>(); 
		for (Map.Entry<String, HashMap<Integer, HashMap<String, String>>> entry : this.campaigns.entrySet())
			if (!campaignsIdList.contains(Integer.parseInt(entry.getKey().split(" ")[0])))
				campaignsIdList.add(Integer.parseInt(entry.getKey().split(" ")[0]));
		return campaignsIdList;
	}
	
	/**
	* @return The list of the names for the specified campaign ID (a campaign can have more than one name).
	*/
	public ArrayList<String> getCampaignNames(int campaignId) {
		ArrayList<String> campaignsName = new ArrayList<String>();
		for (Map.Entry<String, HashMap<Integer, HashMap<String, String>>> entry : this.campaigns.entrySet())
			if (Integer.parseInt(entry.getKey().split(" ")[0]) == campaignId)
				campaignsName.add(entry.getKey().split(" ")[1]);
		return campaignsName;
	}
	
	/**
	* @return The list of all the transactions ID.
	*/
	public ArrayList<Integer> getTransactionsIdList() {
		ArrayList<Integer> transactionsIdList = new ArrayList<Integer>();
		for (Map.Entry<String, HashMap<Integer, HashMap<String, String>>> entry : this.campaigns.entrySet())
			transactionsIdList.addAll(entry.getValue().keySet());
		return transactionsIdList;
	}
	
	/**
	* @return The list of the transactions ID for the specified campaign ID.
	*/
	public ArrayList<Integer> getTransactionsIdListFrom(int campaignId) {
		ArrayList<Integer> transactionsIdList = new ArrayList<Integer>();
		for (Map.Entry<String, HashMap<Integer, HashMap<String, String>>> entry : this.campaigns.entrySet())
			if (Integer.parseInt(entry.getKey().split(" ")[0]) == campaignId)
				transactionsIdList.addAll(entry.getValue().keySet());
		return transactionsIdList;
	}
	
	/**
	* @return The value for the specified key.
	*/
	private String getValue(int campaignId, int transactionId, String key) {
		for (Map.Entry<String, HashMap<Integer, HashMap<String, String>>> entry : this.campaigns.entrySet())
			if (Integer.parseInt(entry.getKey().split(" ")[0]) == campaignId)
				if (entry.getValue().get(transactionId) != null)
					return entry.getValue().get(transactionId).get(key);
		return null;
	}
	
	/**
	* @return The hash corresponding to the specified campaign ID and transaction ID.
	* Null if the campaign ID or the transaction ID does not exist.
	*/
	public String getHash(int campaignId, int transactionId) {
		return getValue(campaignId, transactionId, "hash");
	}
	
	/**
	* @return The datestamp corresponding to the specified campaign ID and transaction ID. 
	* Null if the campaign ID or the transaction ID does not exist.
	*/
	public String getDatestamp(int campaignId, int transactionId) {
		return getValue(campaignId, transactionId, "datestamp");
	}
	
	/**
	* @return The sender's number corresponding to the specified campaign ID and transaction ID.
	* Null if the campaign ID or the transaction ID does not exist.
	*/
	public String getFromNumber(int campaignId, int transactionId) {
		return getValue(campaignId, transactionId, "from_number");
	}
	
	/**
	* @return The sender's name corresponding to the specified campaign ID and transaction ID.
	* Null if the campaign ID or the transaction ID does not exist.
	*/
	public String getFromName(int campaignId, int transactionId) {
		return getValue(campaignId, transactionId, "from_name");
	}
	
	/**
	* @return The receiver's number corresponding to the specified campaign ID and transaction ID.
	* Null if the campaign ID or the transaction ID does not exist.
	*/
	public String getToNumber(int campaignId, int transactionId) {
		return getValue(campaignId, transactionId, "to_number");
	}
	
	/**
	* @return The receiver's name corresponding to the specified campaign ID and transaction ID.
	* Null if the campaign ID or the transaction ID does not exist.
	*/
	public String getToName(int campaignId, int transactionId) {
		return getValue(campaignId, transactionId, "to_name");
	}
}

File: dotNet/MogreetSDK/MogreetSDK/user/Transactions.cs
using System;
using System.Xml;
using System.Security;
using System.Collections.Generic;

namespace MogreetSDK.user
{
    /// <summary>
    /// The Transactions object contains the response from a <see cref="MogreetSDK.Mercury.transactions"/> request to the Moms API.
    /// </summary>
    /// <exception cref='Exception'>
    /// Represents errors that occur during application execution.
    /// </exception>
    public class Transactions : Response
    {
        private Dictionary<string, Dictionary<int, Dictionary<string, string>>> campaigns;

        /// <summary>
        /// Initializes a new instance of the <see cref="SDK.Transactions"/> class.
        /// </summary>
        /// <param name='xmlDoc'>
        /// Xml document that is returned by all the requests.
        /// </param>
        /// <exception cref='Exception'>
        /// Represents errors that occur during application execution.
        /// </exception>
        public Transactions(XmlDocument xmlDoc)
            : base(xmlDoc)
        {
            if (!(base.responseIsValid()))
                throw new Exception(base.getMessage());
            try
            {
                XmlNodeList campList = mXmlDoc.SelectNodes("//response/campaign");
                XmlNodeList transList = mXmlDoc.SelectNodes("//response/campaign/transaction");

                this.campaigns = new Dictionary<string, Dictionary<int, Dictionary<string, string>>>();
                string campaignsIdName = null;
                string datestamp = null;
                string hash = null;
                Dictionary<string, string> transactionsInfo = null;

                for (int i = 0; i < campList.Count; i++)
                {
                    //recovers attributes
                    campaignsIdName = (campList.Item(i).Attributes.Item(0)).Value;
                    campaignsIdName += " " + (campList.Item(i).Attributes.Item(1)).Value;
                    if (!this.campaigns.ContainsKey(campaignsIdName))
                        this.campaigns.Add(campaignsIdName, new Dictionary<int, Dictionary<string, string>>());

                    datestamp = (transList.Item(i).Attributes.Item(0)).Value;
                    hash = (transList.Item(i).Attributes.Item(1)).Value;

                    Dictionary<int, Dictionary<string, string>> transactions = this.campaigns[campaignsIdName];
                    transactions.Add(Int32.Parse((transList.Item(i).Attributes.Item(2)).InnerText), new Dictionary<string, string>()); //messageID

                    transactionsInfo = transactions[Int32.Parse((transList.Item(i).Attributes.Item(2)).InnerText)];
                    transactionsInfo.Add("datestamp", datestamp);
                    transactionsInfo.Add("hash", hash);
                    transactionsInfo.Add("to_number", (transList.Item(i).LastChild.Attributes.Item(0)).Value);
                    transactionsInfo.Add("to_name", transList.Item(i).LastChild.InnerText);
                    transactionsInfo.Add("from_number", (transList.Item(i).FirstChild.Attributes.Item(0)).Value);
                    transactionsInfo.Add("from_name", transList.Item(i).FirstChild.InnerText);
                }
            }
            catch (XmlException e)
            {
                throw new Exception("\nAn error occured while parsing the XML data for the TRANSACTIONS call: " + e.Message);
            }
        }

        /// <summary>
        /// Gets the list of all the campaigns ID.
        /// </summary>
        /// <returns>
        /// The list of all the campaigns ID.
        /// </returns>
        public List<int> getCampaignsIdList()
        {
            List<int> campaignsIdList = new List<int>();
            List<string> keys = new List<string>(this.campaigns.Keys);
            for (int i = 0; i < keys.Count; i++)
            {
                campaignsIdList.Add(Int32.Parse(keys[i].Split(' ')[0]));
            }
            return campaignsIdList;
        }

        /// <summary>
        /// Gets the list of the names for the specified campaign ID (a campaign can have more than one name).
        /// </summary>
        /// <returns>
        /// The list of the names for the specified campaign ID (a campaign can have more than one name).
        /// </returns>
        /// <param name='campaignId'>
        /// Campaign ID.
        /// </param>
        public List<string> getCampaignNames(int campaignId)
        {
            List<string> campaignsName = new List<string>();
            List<string> keys = new List<string>(this.campaigns.Keys);
            int totalLength = 0;
            int idLength = 0;
            int length;
            for (int i = 0; i < keys.Count; i++)
            {
                if (campaignId.Equals(Int32.Parse(keys[i].Split(' ')[0])))
                    totalLength = keys[i].Length;
                idLength = ((keys[i].Split(' ')[0]).Length + 1);
                length = totalLength - idLength;
                campaignsName.Add(keys[i].Substring(idLength, length));
            }
            return campaignsName;
        }

        /// <summary>
        /// Gets the list of the transactions ID for the specified campaign ID.
        /// </summary>
        /// <returns>
        /// The list of the transactions ID for the specified campaign ID.
        /// </returns>
        /// <param name='campaignId'>
        /// Campaign ID.
        /// </param>
        public List<int> getTransactionsIdListFrom(int campaignId)
        {
            List<int> transactionsIdList = new List<int>();
            List<string> keys = new List<string>(this.campaigns.Keys);
            List<int> tmp;
            for (int i = 0; i < keys.Count; i++)
            {
                if (campaignId.Equals(Int32.Parse(keys[i].Split(' ')[0])))
                {
                    tmp = new List<int>(this.campaigns[keys[i]].Keys);
                    foreach (int k in tmp)
                    {
                        transactionsIdList.Add(k);
                    }
                }
            }
            return transactionsIdList;
        }

        /// <summary>
        /// Gets the value for the specified key.
        /// </summary>
        /// <returns>
        /// The value for the specified key.
        /// </returns>
        /// <param name='campaignId'>
        /// Campaign ID.
        /// </param>
        /// <param name='transactionId'>
        /// Transaction ID.
        /// </param>
        /// <param name='key'>
        /// Key.
        /// </param>
        public string getValue(int campaignId, int transactionId, string key)
        {
            List<string> keys = new List<string>(this.campaigns.Keys);
            for (int i = 0; i < keys.Count; i++)
            {
                if (campaignId.Equals(Int32.Parse(keys[i].Split(' ')[0])))
                {
                    if (!(this.campaigns[keys[i]][transactionId].Equals(null)))
                        return this.campaigns[keys[i]][transactionId][key];
                }
            }
            return null;
        }

        /// <summary>
        /// Gets the hash.
        /// </summary>
        /// <returns>
        /// The hash corresponding to the specified campaign ID and transaction ID.
        /// Null if the campaign ID or the transaction ID does not exist.
        /// </returns>
        /// <param name='campaignId'>
        /// Campaign ID.
        /// </param>
        /// <param name='transactionId'>
        /// Transaction ID.
        /// </param>
        public string getHash(int campaignId, int transactionId)
        {
            return getValue(campaignId, transactionId, "hash");
        }

        /// <summary>
        /// Gets the datestamp.
        /// </summary>
        /// <returns>
        /// The datestamp corresponding to the specified campaign ID and transaction ID. 
        /// Null if the campaign ID or the transaction ID does not exist.
        /// </returns>
        /// <param name='campaignId'>
        /// Campaign ID.
        /// </param>
        /// <param name='transactionId'>
        /// Transaction ID.
        /// </param>
        public string getDatestamp(int campaignId, int transactionId)
        {
            return getValue(campaignId, transactionId, "datestamp");
        }

        /// <summary>
        /// Gets the sender's number
        /// </summary>
        /// <returns>
        /// The sender's number corresponding to the specified campaign ID and transaction ID.
        /// Null if the campaign ID or the transaction ID does not exist.
        /// </returns>
        /// <param name='campaignId'>
        /// Campaign ID.
        /// </param>
        /// <param name='transactionId'>
        /// Transaction ID.
        /// </param>
        public string getFromNumber(int campaignId, int transactionId)
        {
            return getValue(campaignId, transactionId, "from_number");
        }

        /// <summary>
        /// Gets the sender's name
        /// </summary>
        /// <returns>
        /// The sender's name corresponding to the specified campaign ID and transaction ID.
        /// Null if the campaign ID or the transaction ID does not exist.
        /// </returns>
        /// <param name='campaignId'>
        /// Campaign ID.
        /// </param>
        /// <param name='transactionId'>
        /// Transaction ID.
        /// </param>
        public string getFromName(int campaignId, int transactionId)
        {
            return getValue(campaignId, transactionId, "from_name");
        }

        /// <summary>
        /// Gets the receiver's number
        /// </summary>
        /// <returns>
        /// The receiver's number corresponding to the specified campaign ID and transaction ID.
        /// Null if the campaign ID or the transaction ID does not exist.
        /// </returns>
        /// <param name='campaignId'>
        /// Campaign ID.
        /// </param>
        /// <param name='transactionId'>
        /// Transaction ID.
        /// </param>
        public string getToNumber(int campaignId, int transactionId)
        {
            return getValue(campaignId, transactionId, "to_number");
        }

        /// <summary>
        /// Gets the receiver's name
        /// </summary>
        /// <returns>
        /// The receiver's name corresponding to the specified campaign ID and transaction ID.
        /// Null if the campaign ID or the transaction ID does not exist.
        /// </returns>
        /// <param name='campaignId'>
        /// Campaign ID.
        /// </param>
        /// <param name='transactionId'>
        /// Transaction ID.
        /// </param>
        public string getToName(int campaignId, int transactionId)
        {
            return getValue(campaignId, transactionId, "to_name");
        }
    }
}
